cpp/src/interop/arrow_data_structures.cpp

```diff
@@ -0,0 +1,345 @@
+    }
+  }
+  output->children     = private_data->children_raw.data();
+  output->dictionary   = input->dictionary;
+  output->release      = ArrayReleaseCallback;
+  output->private_data = private_data;
+}
+
+template <typename T>
+void arrow_obj_to_arrow(T& obj,
+                        std::shared_ptr<arrow_array_container> container,
+                        ArrowDeviceArray* output,
+                        ArrowDeviceType device_type,
+                        rmm::cuda_stream_view stream,
+                        rmm::device_async_resource_ref mr)
+{
+  switch (device_type) {
+    case ARROW_DEVICE_CUDA:
+    case ARROW_DEVICE_CUDA_HOST:
+    case ARROW_DEVICE_CUDA_MANAGED: {
+      auto& device_arr = container->owner;
+      copy_array(&output->array, &device_arr.array, container);
+      output->device_id = device_arr.device_id;
+      // We can reuse the sync event by reference from the input. The
+      // destruction of that event is managed by the destruction of
+      // the underlying ArrowDeviceArray of this table.
+      output->sync_event  = device_arr.sync_event;
+      output->device_type = device_type;
+      break;
+    }
+    case ARROW_DEVICE_CPU: {
+      auto out = cudf::to_arrow_host(*obj.view().get(), stream, mr);
+      ArrowArrayMove(&out->array, &output->array);
+      output->device_id   = -1;
+      output->sync_event  = nullptr;
+      output->device_type = ARROW_DEVICE_CPU;
+      break;
+    }
+    default: throw std::runtime_error("Unsupported ArrowDeviceArray type");
+  }
+}
+
+}  // namespace
+
+arrow_column::arrow_column(cudf::column&& input,
+                           column_metadata const& metadata,
+                           rmm::cuda_stream_view stream,
+                           rmm::device_async_resource_ref mr)
+  : container{[&]() {
+      auto table_meta = std::vector{metadata};
+      auto tv         = cudf::table_view{{input.view()}};
+      auto schema     = cudf::to_arrow_schema(tv, table_meta);
+      return std::make_shared<arrow_array_container>(
+        schema->children[0], std::move(input), stream, mr);
+    }()}
+{
+}
+
+arrow_column::arrow_column(ArrowSchema const* schema,
+                           ArrowDeviceArray* input,
+                           rmm::cuda_stream_view stream,
+                           rmm::device_async_resource_ref mr)
+{
+  switch (input->device_type) {
+    case ARROW_DEVICE_CPU: {
+      auto col        = from_arrow_host_column(schema, input, stream, mr);
+      auto tmp_column = arrow_column(std::move(*col), get_column_metadata(col->view()), stream, mr);
+      container       = tmp_column.container;
+      // Should always be non-null unless we're in some odd multithreaded
+      // context but best to be safe.
+      if (input->array.release != nullptr) { ArrowArrayRelease(&input->array); }
+      break;
+    }
+    default: container = std::make_shared<arrow_array_container>(schema, input, stream, mr);
```

Comment:
you should do an `ArrowArrayMove` from input into what you pass to the `arrow_array_container`. Same for the schema, otherwise the original caller can bypass the `arrow_array_container` and release stuff out from under you.

Reply:
Good point. Yes, we can create a temporary here and move into that before forwarding into the container.

Reply:
The reason I didn't originally do this (and the reason for some of the other minor discrepancies between the two different constructors for the `arrow_array_container`) is because all the `from_arrow*` functions that you previously added accept an `ArrowSchema const*`. In the construction of an `arrow_array_container` from a `cudf::table` or `cudf::column` I just move the schema into the container, but when constructing from an externally provided `ArrowDeviceArray` I currently perform an `ArrowSchemaDeepCopy`  because I matched the `from_arrow_*` signatures and had a const schema. I can change that and then moving things out here will work just fine.

Reply:
Now that I've remove the const bits I think simply forwarding along the arguments is fine since they're all pointers and I'm moving all of them in the arrow_array_container constructors, but let me know if I'm missing anything there.